java/src/main/java/com/microsoft/onnxruntime/ONNXValue.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright © 2019, Oracle and/or its affiliates. All rights reserved.
+ */
+package com.microsoft.onnxruntime;
+
+/**
+ * Top interface for input and output values from ONNX models.
+ * Currently implemented by {@link ONNXTensor}, {@link ONNXSequence} and {@link ONNXMap}. Will be sealed to
+ * these types one day.
+ *
+ * Does not support sparse tensors.
+ */
+public interface ONNXValue extends AutoCloseable {
+
+    /**
+     * The type of the ONNXValue, mirroring the id in the C API.
+     */
+    public enum ONNXValueType {
+        ONNX_TYPE_UNKNOWN(0),
+        ONNX_TYPE_TENSOR(1),
+        ONNX_TYPE_SEQUENCE(2),
+        ONNX_TYPE_MAP(3),
+        ONNX_TYPE_OPAQUE(4),
+        ONNX_TYPE_SPARSETENSOR(5);
+
+        /**
+         * The id number of this type in the C API.
+         */
+        public final int value;
+        ONNXValueType(int value) {
+            this.value = value;
+        }
+    }
+
+    /**
+     * Gets the type of this ONNXValue.
+     * @return The value type.
+     */
+    public ONNXValueType getType();
+
+    /**
+     * Returns the value as a POJO copying it out of the native heap.
+     * This operation can be quite slow for high dimensional tensors.
+     *
+     * Overridden by the subclasses with a more specific type.
+     * @return The value.
+     * @throws ONNXException If an error occurred reading the value.
+     */
+    public Object getValue() throws ONNXException;
+
+    /**
+     * Gets the info object associated with this ONNXValue.
+     * @return The info.
+     */
+    public ValueInfo getInfo();
+
+    /**
+     * Closes the ONNXValue, freeing it's native memory.
+     */
+    @Override
+    public void close();
+
+    /**
+     * Calls close on each element of the iterable.
+     * @param itr An iterable of closeable ONNXValues.
+     */
+    public static void close(Iterable<? extends ONNXValue> itr) {
```

Comment:
> public static void close(Iterable<? extends ONNXValue> itr)  [](start = 3, length = 61)

Would this work in an exception safe manner? Like with try-with_resource?

Reply:
Unfortunately not, there's no way to do that without creating your own list subtype that implements closeable. This static method is a convenience until the Java type system gets more powerful. When specialised generics lands from Valhalla that might allow us to have lists of closeable things be closeable, but I don't think that's settled yet (and it's a few years away).

Reply:
I think creating an class with IterableClosable with AutoClose impl is a fine idea. One then can inherit from it in a generic way only to get a typed handle of the actual collection. You could then use it with try-with-resource.

---
In reply to: [340888190](https://github.com/microsoft/onnxruntime/pull/2215#discussion_r340888190) [](ancestors = 340888190)

Reply:
It's not possible to do the type mixin onto an existing concrete collection instance, I'd need to subclass ArrayList and have it implement AutoCloseable. I'm wary of creating a list type that's only used by this project. It's likely to cause issues if people start using it elsewhere in their projects.

Reply:
I think the above needs to be reflected in the comments along with the suggested use of the method. I personally, would opt to create a class as a scoped guard. It seems better to have something than nothing. You can have your collection and then you can have IterableClosable with that collection in the constructor within try-with-resource statement. I think you  can even pass Collection interface there instead of Iterable. Without it I do not know how I would write code right this moment.

---
In reply to: [341117653](https://github.com/microsoft/onnxruntime/pull/2215#discussion_r341117653) [](ancestors = 341117653)

Reply:
Where you have a call to `close` on an `ONNXValue` you can replace that with a call to `ONNXValue.close` supplying a list of ONNXValues. I agree it's a bad name, it should be renamed to something like `closeIterable` or `closeAll`, and I can rework the javadoc. Iterable is a more general interface than collection, so I chose that. I can narrow the type if you want.

I'm not sure what you mean by "create a class as a scoped guard"? The `session.score` method returns a `List<ONNXValue>`, and so this method is the convenience method to help clear up the output of an inference run. That output could be a subtype of List that also implements Closeable, but that would be a specific subtype to this project, which is kinda weird. I suppose the implementation of the interface could be private, but I'd still be worried about having a private subclass of ArrayList in any project.

Reply:
@Craigacp I think the point is to have something that implements `AutoCloseable`, like the `TensorScope` we've been discussing for TensorFlow that you don't like. :) Then we can use try-with-resources statements to handle magically all exceptions correctly.